File: echolot-demo/src/main/java/de/exxcellent/echolot/testapp/testscreen/LineChartTest.java
/*
 * This file (LineChartTest.java) is part of the Echolot Project (hereinafter "Echolot").
 * Copyright (C) 2008-2010 eXXcellent Solutions GmbH.
 *
 * Version: MPL 1.1/GPL 2.0/LGPL 2.1
 *
 * The contents of this file are subject to the Mozilla Public License Version
 * 1.1 (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 * http://www.mozilla.org/MPL/
 *
 * Software distributed under the License is distributed on an "AS IS" basis,
 * WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License
 * for the specific language governing rights and limitations under the
 * License.
 *
 * Alternatively, the contents of this file may be used under the terms of
 * either the GNU General Public License Version 2 or later (the "GPL"), or
 * the GNU Lesser General Public License Version 2.1 or later (the "LGPL"),
 * in which case the provisions of the GPL or the LGPL are applicable instead
 * of those above. If you wish to allow use of your version of this file only
 * under the terms of either the GPL or the LGPL, and not to allow others to
 * use your version of this file under the terms of the MPL, indicate your
 * decision by deleting the provisions above and replace them with the notice
 * and other provisions required by the GPL or the LGPL. If you do not delete
 * the provisions above, a recipient may use your version of this file under
 * the terms of any one of the MPL, the GPL or the LGPL.
 */

package de.exxcellent.echolot.testapp.testscreen;

import de.exxcellent.echolot.app.LineChart;
import de.exxcellent.echolot.event.PointSelectEvent;
import de.exxcellent.echolot.listener.PointSelectListener;
import de.exxcellent.echolot.model.AxisModel;
import de.exxcellent.echolot.model.Interpolation;
import de.exxcellent.echolot.model.LineChartModel;
import de.exxcellent.echolot.model.LineChartPoint;
import de.exxcellent.echolot.testapp.ButtonColumn;
import de.exxcellent.echolot.testapp.StyleUtil;
import nextapp.echo.app.Color;
import nextapp.echo.app.Extent;
import nextapp.echo.app.Label;
import nextapp.echo.app.SplitPane;
import nextapp.echo.app.event.ActionEvent;
import nextapp.echo.app.event.ActionListener;

public class LineChartTest extends SplitPane {

    private LineChart lineChart;

    /**
     * Visual Test for the Pie-Chart
     */
    public LineChartTest() {
        super(SplitPane.ORIENTATION_HORIZONTAL, new Extent(250, Extent.PX));
        setStyleName("DefaultResizable");

        ButtonColumn controlsColumn = new ButtonColumn();
        controlsColumn.setStyleName("TestControlsColumn");
        add(controlsColumn);

        // Default
        lineChart = new LineChart();
        //setLineChart_1();
        //setLineChartModel_1();      

        setLineChart_2();
        setLineChartModel_2();

        // Listener
        lineChart.addPointSelectListener(new PointSelectListener() {
            public void pointSelection(PointSelectEvent pointSelectEvent) {
                System.out.println("You clicked Point with Label: " + pointSelectEvent.getLineChartPoint().getLabel() + " \n X/Y values::  X-Value: " + pointSelectEvent.getLineChartPoint().getxValue() + " Y-Value: " + pointSelectEvent.getLineChartPoint().getyValue());
            }
        });

        add(lineChart);


        // -- Models ---
        controlsColumn.add(new Label("Different LineChartModel"));

        controlsColumn.addButton("LineChart == 'null'", new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                setLineChartModelToNull();
            }
        });

        controlsColumn.addButton("Default LineChart", new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                setLineChart_1();
                setLineChartModel_1();
            }
        });
        controlsColumn.addButton("LineChart with more Nodes", new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                setLineChart_2();
                setLineChartModel_2();
            }
        });


        // Layout
        controlsColumn.add(new Label("Common Layout"));
        controlsColumn.addButton("width +10", new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                changeWidth(+10);
            }
        });
        controlsColumn.addButton("width -10", new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                changeWidth(-10);
            }
        });
        controlsColumn.addButton("height +10", new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                changeHeight(+10);
            }
        });
        controlsColumn.addButton("height -10", new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                changeHeight(-10);
            }
        });


        // Scaling
        controlsColumn.add(new Label("Different Scaling"));
        controlsColumn.addButton("xMax +10", new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                changeXscaleMax(+10);
            }
        });
        controlsColumn.addButton("xMax -10", new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                changeXscaleMax(-10);
            }
        });
        controlsColumn.addButton("yMax +10", new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                changeYscaleMax(+10);
            }
        });
        controlsColumn.addButton("yMax -10", new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                changeYscaleMax(-10);
            }
        });

        // Axis
        controlsColumn.add(new Label("Different Scaling"));
        controlsColumn.addButton("AxisModel Calender", new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                setAxisModel();
            }
        });

        controlsColumn.addButton("AxisModel NULL", new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                setAxisModelNULL();
            }
        });

        controlsColumn.addButton("Random AxisForegroundColor", new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                setRandomAxisForegroundColor();
            }
        });


        // Grid
        controlsColumn.add(new Label("Grid options"));
        controlsColumn.addButton("Toggle Grid on/off", new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                toogleGridOnOff();
            }
        });
        controlsColumn.addButton("# xAxis Sector +1", new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                changeXaxisSectors(+1);
            }
        });
        controlsColumn.addButton("# xAxis Sector -1", new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                changeXaxisSectors(-1);
            }
        });
        controlsColumn.addButton("# yAxis Sector +1", new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                changeYaxisSectors(+1);
            }
        });
        controlsColumn.addButton("# yAxis Sector -1", new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                changeYaxisSectors(-1);
            }
        });
        controlsColumn.addButton("Random Gridcolor", new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                setRandomGridColor();
            }
        });

        // Chart-Options
        controlsColumn.add(new Label("Line options"));
        controlsColumn.addButton("Toggle Popup on/off", new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                tooglePopupOnOff();
            }
        });

        controlsColumn.addButton("Toggle FillChart on/off", new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                toogleFillOnOff();
            }
        });

        controlsColumn.addButton("Random Linecolor", new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                setRandomLineColor();
            }
        });

        controlsColumn.addButton("Random Dotcolor", new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                setRandomDotColor();
            }
        });

        // Popup-Styling
        controlsColumn.add(new Label("Popup-Styling"));
        controlsColumn.addButton("Random border-color", new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                setRandomBorderColor();
            }
        });

        controlsColumn.addButton("Random background-color", new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                setRandomBackgroundColor();
            }
        });

        controlsColumn.addButton("Random foreground-color", new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                setRandomForegroundColor();
            }
        });

        controlsColumn.addButton("Random font", new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                setRandomFont();
            }
        });

        // Interpolation
        controlsColumn.add(new Label("Interpolation"));
        controlsColumn.addButton("Interpolation: NONE", new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                setInterpolation(Interpolation.NONE);
            }
        });
        controlsColumn.addButton("Interpolation: LINEAR", new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                setInterpolation(Interpolation.LINEAR);
            }
        });
        controlsColumn.addButton("Interpolation: BEZIER", new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                setInterpolation(Interpolation.BEZIER);
            }
        });

    }


    private void setLineChart_1() {
        lineChart.setWidth(300);
        lineChart.setHeight(200);
        lineChart.setDrawGrid(true);

        lineChart.setXscaleMax(100);
        lineChart.setYscaleMax(100);

        lineChart.setXaxisSectors(10);
        lineChart.setYaxisSectors(10);

        lineChart.setShowPopup(true);
        lineChart.setPopupBackground(Color.DARKGRAY);
        lineChart.setPopupBorderColor(Color.LIGHTGRAY);
        lineChart.setPopupForeground(Color.WHITE);
        lineChart.setPopupFont(StyleUtil.randomFont());

        //lineChart.setDotColor("#383");
        lineChart.setDotColor(Color.BLUE);
        lineChart.setLineColor(Color.DARKGRAY);

        lineChart.setFillChart(true);
    }

    private void setLineChart_2() {
        lineChart.setWidth(450);
        lineChart.setHeight(300);

        lineChart.setDrawGrid(true);

        lineChart.setXscaleMax(200);
        lineChart.setYscaleMax(150);

        lineChart.setXaxisSectors(20);
        lineChart.setYaxisSectors(15);

        lineChart.setShowPopup(true);

        lineChart.setDotColor(Color.BLUE);
        lineChart.setLineColor(Color.DARKGRAY);

        lineChart.setFillChart(true);
    }

    private void setLineChartModelToNull() {
        lineChart.setLineChartModel(null);
    }

    private void setLineChartModel_1() {
        LineChartPoint point_1 = new LineChartPoint(0, 20, "Tolles Label");
        LineChartPoint point_2 = new LineChartPoint(40, 80, "Tolles Label");
        LineChartPoint point_3 = new LineChartPoint(60, 30, "Tolles Label");
        LineChartPoint point_4 = new LineChartPoint(100, 50, "Tolles Label");

        LineChartPoint[] pointArray = new LineChartPoint[]{
                point_1,
                point_2,
                point_3,
                point_4
        };

        lineChart.setLineChartModel(new LineChartModel(pointArray));
    }

    private void setLineChartModel_2() {
        LineChartPoint point_1 = new LineChartPoint(0, 20, "Tolles Label 1 \n 0/20");
        LineChartPoint point_2 = new LineChartPoint(10, 60, "Tolles Label 2 \n 10/60");
        LineChartPoint point_3 = new LineChartPoint(15, 50, "Tolles Label 3 \n 15/50");
        LineChartPoint point_4 = new LineChartPoint(25, 60, "Tolles Label 4 \n 25/60");
        LineChartPoint point_5 = new LineChartPoint(40, 80, "Tolles Label 5 \n 40/80");
        LineChartPoint point_6 = new LineChartPoint(55, 20, "Tolles Label 6 \n 55/20");
        LineChartPoint point_7 = new LineChartPoint(70, 85, "Tolles Label 7 \n 70/85");
        LineChartPoint point_8 = new LineChartPoint(80, 60, "Tolles Label 8 \n 80/60");
        LineChartPoint point_9 = new LineChartPoint(100, 50, "Tolles Label 9 \n 100/50");
        LineChartPoint point_10 = new LineChartPoint(120, 70, "Tolles Label 10 \n 120/70");
        LineChartPoint point_11 = new LineChartPoint(125, 30, "Tolles Label 11 \n 125/30");
        LineChartPoint point_12 = new LineChartPoint(140, 45, "Tolles Label 12 \n 140/45");
        LineChartPoint point_13 = new LineChartPoint(150, 80, "Tolles Label 13 \n 150/80");
        LineChartPoint point_14 = new LineChartPoint(190, 50, "Tolles Label 14 \n 190/50");
        LineChartPoint point_15 = new LineChartPoint(200, 70, "Tolles Label 15 \n 200/70");

        LineChartPoint[] pointArray = new LineChartPoint[]{
                point_1,
                point_2,
                point_3,
                point_4,
                point_5,
                point_6,
                point_7,
                point_8,
                point_9,
                point_10,
                point_11,
                point_12,
                point_13,
                point_14,
                point_15
        };

        lineChart.setLineChartModel(new LineChartModel(pointArray));
    }

    private void setAxisModel() {
        String[] xAxisValues = new String[]{"Jan", "Feb", "Mrz", "Apr"};
        String[] yAxisValues = new String[]{"a", "b", "c"};

        lineChart.setAxisModel(new AxisModel(xAxisValues, yAxisValues));
    }

    private void setAxisModelNULL() {
        lineChart.setAxisModel(null);
    }

    private void changeWidth(int factor) {
        lineChart.setWidth(lineChart.getWidth() + factor);
    }

    private void changeHeight(int factor) {
        lineChart.setHeight(lineChart.getHeight() + factor);
    }

    private void changeXscaleMax(int factor) {
        lineChart.setXscaleMax(lineChart.getXscaleMax() + factor);
    }

    private void changeYscaleMax(int factor) {
        lineChart.setYscaleMax(lineChart.getYscaleMax() + factor);
    }

    // axisOptions

    private void setRandomAxisForegroundColor() {
        lineChart.setForeground(StyleUtil.randomColor());
    }

    // grid

    private void toogleGridOnOff() {
        lineChart.setDrawGrid(!lineChart.isDrawGrid());
    }

    private void changeXaxisSectors(int factor) {
        lineChart.setXaxisSectors(lineChart.getXaxisSectors() + factor);
    }

    private void changeYaxisSectors(int factor) {
        lineChart.setYaxisSectors(lineChart.getYaxisSectors() + factor);
    }

    private void setRandomGridColor() {
        lineChart.setGridColor(StyleUtil.randomColor());
    }

    // line

    private void tooglePopupOnOff() {
        lineChart.setShowPopup(!lineChart.isShowPopup());
    }

    private void toogleFillOnOff() {
        lineChart.setFillChart(!lineChart.isFillChart());
    }

    private void setRandomLineColor() {
        lineChart.setLineColor(StyleUtil.randomBrightColor());
    }

    private void setRandomDotColor() {
        lineChart.setDotColor(StyleUtil.randomColor());
    }

    private void setInterpolation(Interpolation interpolation) {
        lineChart.setInterpolation(interpolation);
    }

    private void setRandomBorderColor() {
        lineChart.setPopupBorderColor(StyleUtil.randomBrightColor());
    }

    private void setRandomBackgroundColor() {
        lineChart.setPopupBackground(StyleUtil.randomBrightColor());
    }

    private void setRandomForegroundColor() {
        lineChart.setPopupForeground(StyleUtil.randomBrightColor());
    }

    private void setRandomFont() {
        lineChart.setPopupFont(StyleUtil.randomFont());
    }

    private String getRandomColor() {
        String color = "#";
        for (int i = 0; i < 3; i++) {
            String hexColor = Integer.toHexString((int) (Math.random() * 16));
            color = color + hexColor;
        }
        return color;
    }


}

File: echolot-webcontainer/src/test/web/testscreen/LineChartTest.js
/*
 * This file (LineChartTest.js) is part of the Echolot Project (hereinafter "Echolot").
 * Copyright (C) 2008-2011 eXXcellent Solutions GmbH.
 *
 * Version: MPL 1.1/GPL 2.0/LGPL 2.1
 */

/**
 * Testclass for the LineChart
 *
 * @author Ralf Enderle <r.enderle@exxcellent.de>
 */
exxcellent.test.LineChartTest = Core.extend({
    _lineChart: null,

    $construct: function (testArea, mainContent) {
        var controlColumn = new Echo.Column({
            styleName: "TestControl",
            children: [
                new Echo.Label({
                    text: "LineChart", styleName: "Title"
                }),
                new Echo.Label({
                    text: "Models", styleName: "Default"
                }),
                this._createButton("LineChart Model 1",
                    function () {
                        this._lineChart.set(exxcellent.LineChart.LINE_CHART_MODEL, this._createLineChartModel_1().lineChartModel);
                    }
                ),
                this._createButton("LineChart Model 2",
                    function () {
                        this._lineChart.set(exxcellent.LineChart.LINE_CHART_MODEL, this._createLineChartModel_2().lineChartModel);
                    }
                ),
                this._createButton("LineChart Model 3",
                    function () {
                        this._lineChart.set(exxcellent.LineChart.LINE_CHART_MODEL, this._createLineChartModel_3().lineChartModel);
                    }
                ),
                this._createButton("Axis Model 1",
                    function () {
                        this._lineChart.set(exxcellent.LineChart.AXIS_MODEL, this._createAxisModel_1().axisModel);
                    }
                ),
                this._createButton("Toggle - showPopup",
                    function () {
                        this._lineChart.set(exxcellent.LineChart.SHOW_POPUP, !this._lineChart.get(exxcellent.LineChart.SHOW_POPUP));
                    }),
                this._createButton("Toggle - fillChart",
                    function () {
                        this._lineChart.set(exxcellent.LineChart.FILL_CHART, !this._lineChart.get(exxcellent.LineChart.FILL_CHART));
                    }),
                this._createButton("Toggle - drawGrid",
                    function () {
                        this._lineChart.set(exxcellent.LineChart.DRAW_GRID, !this._lineChart.get(exxcellent.LineChart.DRAW_GRID));
                    }),
                new Echo.Label({
                    text: "Styling", styleName: "Default"
                }),
                this._createButton("Set Random Foreground",
                    function () {
                        this._lineChart.set(exxcellent.LineChart.FOREGROUND, Raphael.getColor());
                    }),
                this._createButton("Set Random LineColor",
                    function () {
                        this._lineChart.set(exxcellent.LineChart.LINE_COLOR, Raphael.getColor());
                    }),
                this._createButton("Set Random DotColor",
                    function () {
                        this._lineChart.set(exxcellent.LineChart.DOT_COLOR, Raphael.getColor());
                    }),
                this._createButton("Set Random GridColor",
                    function () {
                        this._lineChart.set(exxcellent.LineChart.GRID_COLOR, Raphael.getColor());
                    }),
                new Echo.Label({text: "Line-Interpolation", styleName: "Default"}),
                this._createButton("NONE",
                    function () {
                        this._lineChart.set(exxcellent.LineChart.INTERPOLATION, 'none');
                    }),
                this._createButton("LINEAR",
                    function () {
                        this._lineChart.set(exxcellent.LineChart.INTERPOLATION, 'linear');
                    }),
                this._createButton("BEZIER",
                    function () {
                        this._lineChart.set(exxcellent.LineChart.INTERPOLATION, 'bezier');
                    }),
                new Echo.Label({
                    text: "Scaling", styleName: "Default"
                }),
                this._createButton("xScaleMax + 10",
                    function () {
                        this._lineChart.set(exxcellent.LineChart.XSCALE_MAX, this._lineChart.get(exxcellent.LineChart.XSCALE_MAX) + 10);
                    }),
                this._createButton("xScaleMax - 10",
                    function () {
                        this._lineChart.set(exxcellent.LineChart.XSCALE_MAX, this._lineChart.get(exxcellent.LineChart.XSCALE_MAX) - 10);
                    }),
                this._createButton("xSectors + 1",
                    function () {
                        this._lineChart.set(exxcellent.LineChart.XAXIS_SECTORS, this._lineChart.get(exxcellent.LineChart.XAXIS_SECTORS) + 1);
                    }),
                this._createButton("xSectors - 1",
                    function () {
                        this._lineChart.set(exxcellent.LineChart.XAXIS_SECTORS, this._lineChart.get(exxcellent.LineChart.XAXIS_SECTORS) - 1);
                    }),
                this._createButton("yScaleMax + 10",
                    function () {
                        this._lineChart.set(exxcellent.LineChart.YSCALE_MAX, this._lineChart.get(exxcellent.LineChart.YSCALE_MAX) + 10);
                    }),
                this._createButton("yScaleMax - 10",
                    function () {
                        this._lineChart.set(exxcellent.LineChart.YSCALE_MAX, this._lineChart.get(exxcellent.LineChart.YSCALE_MAX) - 10);
                    }),
                this._createButton("ySectors + 1",
                    function () {
                        this._lineChart.set(exxcellent.LineChart.YAXIS_SECTORS, this._lineChart.get(exxcellent.LineChart.YAXIS_SECTORS) + 1);
                    }),
                this._createButton("ySectors - 1",
                    function () {
                        this._lineChart.set(exxcellent.LineChart.YAXIS_SECTORS, this._lineChart.get(exxcellent.LineChart.YAXIS_SECTORS) - 1);
                    }),
                new Echo.Label({
                    text: "Size", styleName: "Default"
                }),
                this._createButton("width + 10",
                    function () {
                        this._lineChart.set(exxcellent.LineChart.WIDTH, this._lineChart.get(exxcellent.LineChart.WIDTH) + 10);
                    }),
                this._createButton("width - 10",
                    function () {
                        this._lineChart.set(exxcellent.LineChart.WIDTH, this._lineChart.get(exxcellent.LineChart.WIDTH) - 10);
                    }),
                this._createButton("height + 10",
                    function () {
                        this._lineChart.set(exxcellent.LineChart.HEIGHT, this._lineChart.get(exxcellent.LineChart.HEIGHT) + 10);
                    }),
                this._createButton("height - 10",
                    function () {
                        this._lineChart.set(exxcellent.LineChart.HEIGHT, this._lineChart.get(exxcellent.LineChart.HEIGHT) - 10);
                    })
            ]
        });

        this._lineChart = this._createLineChart();
        this._lineChart.set(exxcellent.LineChart.LINE_CHART_MODEL, this._createLineChartModel_1().lineChartModel);

        var splitPane = new Echo.SplitPane({
            styleName: "TestControl",
            children: [
                controlColumn,
                new Echo.Column({
                    styleName: "Default",
                    children: [this._lineChart]
                })
            ]
        });

        testArea.add(splitPane);
    },

    _createLineChart: function () {
        return new exxcellent.LineChart({
            font: {
                size: "12",
                typeface: ["Courier New", "Courier", "Monospace"],
                italic: true
            },
            background: '#ffcc66',
            drawGrid: true,
            gridColor: '#000',
            xaxisSectors: 10,
            yaxisSectors: 10,
            showPopup: true,
            popupBackground: '#6699ff',
            popupBorderColor: '#ff6666',
            popupForeground: '#ffffcc',
            popupFont: {
                size: "10",
                typeface: ["Courier New", "Courier", "Monospace"],
                italic: true
            },
            fillChart: true,
            lineColor: '#383',
            dotColor: '#000',
            xscaleMax: 100,
            yscaleMax: 100,
            width: 300,
            height: 200
        });
    },

    /**
     * Creates a LineChartModel
     */
    _createLineChartModel_1: function () {
        // the sectors
        var points = [
            new exxcellent.model.Point(0, 20, 'Label-Text 1', 1),
            new exxcellent.model.Point(10, 80, 'Label-Text 2', 2),
            new exxcellent.model.Point(25, 30, 'Label-Text 3', 3),
            new exxcellent.model.Point(50, 90, 'Label-Text 4', 4),
            new exxcellent.model.Point(70, 10, 'Label-Text 5', 5),
            new exxcellent.model.Point(90, 25, 'Label-Text 6', 6),
            new exxcellent.model.Point(100, 0, 'Label-Text 7', 7)
        ];

        // return the whole thing
        return {
            lineChartModel: new exxcellent.model.LineChartModel(points)
        };
    },
    /**
     * Creates a LineChartModel
     */
    _createLineChartModel_2: function () {
        // the sectors
        var points = [
            new exxcellent.model.Point(0, 50, 'Punkt 0/50', 1),
            new exxcellent.model.Point(20, 80, 'Punkt 20/80', 2),
            new exxcellent.model.Point(70, 70, 'Punkt 70/70', 3),
            new exxcellent.model.Point(100, 70, 'Punkt 100/70', 4)
        ];

        // return the whole thing
        return {
            lineChartModel: new exxcellent.model.LineChartModel(points)
        };
    },
    /**
     * Creates a LineChartModel
     */
    _createLineChartModel_3: function () {
        // the sectors
        var points = [
            new exxcellent.model.Point(0, 20, '0/20', 1),
            new exxcellent.model.Point(10, 80, '10/80', 2),
            new exxcellent.model.Point(25, 30, '25/30', 3),
            new exxcellent.model.Point(35, 50, '35/50', 4),
            new exxcellent.model.Point(45, 20, '45/20', 5),
            new exxcellent.model.Point(48, 30, '48/30', 6),
            new exxcellent.model.Point(55, 70, '55/70', 7),
            new exxcellent.model.Point(60, 30, '60/30', 8),
            new exxcellent.model.Point(70, 10, '70/10', 8),
            new exxcellent.model.Point(90, 25, '90/25', 9),
            new exxcellent.model.Point(100, 0, '100/0', 10)
        ];

        // return the whole thing
        return {
            lineChartModel: new exxcellent.model.LineChartModel(points)
        };
    },

    /**
     * Creates a LineChartModel
     */
    _createAxisModel_1: function () {
        // the sectors
        var xAxisValues = [
            'xTest 1',
            'xTest 2',
            'xTest 3',
            'xTest 4'
        ];
        var yAxisValues = [
            'yTest 1',
            'yTest 2',
            'yTest 3'
        ];

        // return the whole thing
        return {
            axisModel: new exxcellent.model.AxisModel(xAxisValues, yAxisValues)
        };
    },

    _createButton: function (text, action) {
        return new Echo.Button({
            text: text,
            styleName: "Default",
            events: {
                action: Core.method(this, action)
            }
        });
    }
});